xdsl/transforms/loop_nest_lowering_utils.py

```diff
@@ -68,6 +62,105 @@ def indices_for_map(
     return output_indices
 
 
+def _insert_loop_nest(
+    rewriter: PatternRewriter,
+    insertion_point: InsertPoint,
+    zero_op: arith.Constant,
+    one_op: arith.Constant,
+    bounds: tuple[OpResult, ...],
+    iter_args: Sequence[SSAValue],
+    make_body: Callable[
+        [PatternRewriter, InsertPoint, Sequence[BlockArgument], Sequence[SSAValue]],
+        Sequence[SSAValue],
+    ],
+) -> Sequence[SSAValue]:
+    """
+    Creates a perfect loop nest, populating the innermost body with the provided
+    `make_body` function.
+    If `iter_args` are passed in, the loop nest will pass them from parent loop to child
+    loop, and the results of `make_body` are expected to be equal in length to
+    `iter_args`.
+    """
+    if not bounds:
+        return make_body(rewriter, insertion_point, (), iter_args)
+
+    iter_arg_types = tuple(arg.type for arg in iter_args)
+    loops: list[scf.For] = []
+    index = IndexType()
+
+    for i, ub in enumerate(bounds):
+        loop = scf.For(
+            zero_op.result,
+            ub,
+            one_op.result,
+            iter_args,
+            Region(Block(arg_types=(index, *iter_arg_types))),
+        )
+        iter_args = loop.body.block.args[1:]
+        loops.append(loop)
+        rewriter.insert_op_at_location(loop, insertion_point)
+        results = loop.results
+
+        if i + 1 == len(bounds):
```

Comment:
What is this line trying to check?

Reply:
that it's the last iteration in the loop

Reply:
ok, sorry I thought by default enumerate starts at 1 for a moemnt.